### PR TITLE
Refactor Route Detail VC

### PR DESCRIPTION
**Pull request by Omarrasheed**
#238 

Instead of storing bus stops as empty direction objects inserted into our vc's `self.directions` array, we now have a new enum `RouteDetailItem` which is either a BusStop (aka Location Object) or Direction, and our directions array is now of type `[RouteDetailItem]`

### CODE REVIEW

**Review by young-k — APPROVED** [review]
LGTM, but you should probably wait for someone with more iOS knowledge

File: TCAT/Controllers/RouteDetail+ContentViewController.swift [hunk 1]
```diff
@@ -32,9 +29,11 @@ class RouteDetailContentViewController: UIViewController, GMSMapViewDelegate, CL
     private var liveTrackingNetworkRefreshRate: Double = LiveIndicator.INTERVAL * 1.0
     private var liveTrackingNetworkTimer: Timer?
     private var locationManager = CLLocationManager()
+    private var mapView: GMSMapView!
     private let networking: Networking = URLSession.shared.request
     private var paths: [Path] = []
     private var route: Route!
+    private var routeOptionsCell: RouteTableViewCell?
```
Comment (kevinchan159) [inline TCAT/Controllers/RouteDetail+ContentViewController.swift:36]: why do we need to store a reference to this?

File: TCAT/Controllers/RouteDetail+ContentViewController.swift [hunk 2]
```diff
@@ -741,4 +510,178 @@ class RouteDetailContentViewController: UIViewController, GMSMapViewDelegate, CL
...
+    }
+
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+
+        // Live Tracking Network Timer
+        liveTrackingNetworkTimer?.invalidate()
+        if directions.contains(where: { $0.type != .walk }) {
+            liveTrackingNetworkTimer = Timer.scheduledTimer(timeInterval: liveTrackingNetworkRefreshRate, target: self, selector: #selector(getBusLocations),
+                                                            userInfo: nil, repeats: true)
```
Comment (kevinchan159) [inline TCAT/Controllers/RouteDetail+ContentViewController.swift]: if we're splitting this into two lines, we should have each arg on its own line
```
liveTrackingNetworkTimer = Timer.scheduledTimer(
  timeInterval: liveTrackingNetworkRefreshRate, 
  target: self, 
  selector: #selector(getBusLocations),
  userInfo: nil,
  repeats: true
)
```

File: TCAT/Controllers/RouteDetail+ContentViewController.swift [hunk 3]
```diff
@@ -741,4 +510,178 @@ class RouteDetailContentViewController: UIViewController, GMSMapViewDelegate, CL
...
+    }
+
+    override func viewWillDisappear(_ animated: Bool) {
+        super.viewWillDisappear(animated)
+        liveTrackingNetworkTimer?.invalidate()
+        hideBanner()
+    }
+
+    override func loadView() {
+
```
Comment (kevinchan159) [inline TCAT/Controllers/RouteDetail+ContentViewController.swift]: delete newline

File: TCAT/Controllers/RouteDetail+ContentViewController.swift [hunk 4]
```diff
@@ -741,4 +510,178 @@ class RouteDetailContentViewController: UIViewController, GMSMapViewDelegate, CL
...
+extension RouteDetailContentViewController: GMSMapViewDelegate {
+
+    func mapView(_ mapView: GMSMapView, didTap marker: GMSMarker) -> Bool {
+        guard let coordinates = getUserData(for: marker, key: Constants.BusUserData.actualCoordinates) as? CLLocationCoordinate2D
+            else { return true }
+        let update = GMSCameraUpdate.setTarget(coordinates)
+        mapView.animate(with: update)
+
+        return true
+
```
Comment (kevinchan159) [inline TCAT/Controllers/RouteDetail+ContentViewController.swift]: delete newline

File: TCAT/Controllers/RouteDetail+ContentViewController.swift [hunk 5]
```diff
@@ -741,4 +510,178 @@ class RouteDetailContentViewController: UIViewController, GMSMapViewDelegate, CL
...
+    }
+
+    func updateUserData(for marker: GMSMarker, with values: [String: Any]) {
+        guard var userData = marker.userData as? [String: Any] else {
+            marker.userData = values
+            return
+        }
+        for (key, value) in values {
```
Comment (kevinchan159) [inline TCAT/Controllers/RouteDetail+ContentViewController.swift]: pref `forEach`

File: TCAT/Controllers/RouteDetail+ContentViewController.swift [hunk 6]
```diff
@@ -741,4 +510,178 @@ class RouteDetailContentViewController: UIViewController, GMSMapViewDelegate, CL
...
+            userData[key] = value
+        }
+        marker.userData = userData
+    }
+
+    func getUserData(for marker: GMSMarker, key: String) -> Any? {
+        guard var userData = marker.userData as? [String: Any] else { return nil }
```
Comment (kevinchan159) [inline TCAT/Controllers/RouteDetail+ContentViewController.swift]: does it have to be `var userData`? can it be `let`

File: TCAT/Controllers/RouteDetail+ContentViewController.swift [hunk 7]
```diff
@@ -741,4 +510,178 @@ class RouteDetailContentViewController: UIViewController, GMSMapViewDelegate, CL
...
+        return userData[key]
+    }
+
+    func mapView(_ mapView: GMSMapView, didChange position: GMSCameraPosition) {
+        for bus in buses {
```
Comment (kevinchan159) [inline TCAT/Controllers/RouteDetail+ContentViewController.swift]: pref `forEach`

File: TCAT/Controllers/RouteDetail+DrawerViewController.swift [hunk 8]
```diff
@@ -18,37 +18,52 @@ struct RouteDetailCellSize {
...
         super.init(nibName: nil, bundle: nil)
         self.route = route
-        self.directions = route.directions
+        self.directionsAndVisibleStops = route.directions.map({ RouteDetailItem.direction($0) })
     }
 
     func update(with route: Route) {
         self.route = route
-        self.directions = route.directions
+        self.directionsAndVisibleStops = route.directions.map({ RouteDetailItem.direction($0) })
```
Comment (kevinchan159) [inline TCAT/Controllers/RouteDetail+DrawerViewController.swift:66]: cleannn!

**Review by kevinchan159 — CHANGES_REQUESTED** [review]
amazing!!

File: TCAT/Controllers/RouteDetail+ContentViewController.swift (hunk 1 above)
Comment (Omarrasheed) [reply TCAT/Controllers/RouteDetail+ContentViewController.swift:36]: It was being used to take a screenshot of the cell when a user shares but I'll make that a delegate function n just handle that in the main vc

Comment (Omarrasheed) [reply TCAT/Controllers/RouteDetail+ContentViewController.swift:36]: Wait actually I'm not sure if we can do that. Cuz it takes an image of the route options cell which is like an overview of the route and then the user can share that specific route with someone else within the route details page. I'll think of a better way to do this but I'll probably leave this for now if that's ok

**Comment by mattbarker016** [comment]
Reviewed 👍

**Pull request merged**